nixos/modules/system/activation/switch-to-configuration.pl

```diff
@@ -229,7 +229,17 @@ sub fingerprintUnit {
                 # Reload the changed mount unit to force a remount.
                 $unitsToReload{$unit} = 1;
                 recordUnit($reloadListFile, $unit);
-            } elsif ($unit =~ /\.socket$/ || $unit =~ /\.path$/ || $unit =~ /\.slice$/) {
+            } elsif ($unit =~ /\.socket$/) {
+                my $unitInfo = parseUnit($newUnitFile);
+                # If a socket unit has been changed, the corresponding
+                # service unit has to be stopped before the socket can
+                # be restarted.
+                my $serviceUnit = $unitInfo->{'Unit'} // "$baseName.service";
```

Comment:
What happens for services like openssh (when socket activation is turned on), where `sshd.socket` sets `Accept=true` and so sshd instances are spawned from the `sshd@.service` template, but there is no single `sshd.service` running?
IMO we should make sure to keep the [current behavior](https://github.com/NixOS/nixpkgs/issues/43341) that helps avoid accidental lockouts:
- when socket activation is not on, the running (single) sshd is not restarted to avoid potentially dropping the connection used to run `switch-to-configuration`
- with socket activation on, new connections will reflect the new configuration but existing connections are left alone

I'd appreciate some comments here, and perhaps a NixOS test.

Reply:
OpenSSH always forks every new connection into new process that is not part of the main processes process group so systemd will not kill SSH connections. This should be safe and was before the regression mentioned was introduced that made it not restart on config changes.

In case of socket-activation, there is a `ssh@` meta-service whose instances won't be restarted with these changes.

I will definitely add another test for OpenSSh though. Better safe than sorry.